Flexp.setup doesn't clean the experiment directory.

When creating an experiment, I usually did clean_experiment_folder to get rid of previous results. 
Now if I do this:
```python
    flexp.setup(EXPS_DIR, DESC, with_date=not SHORT_RUN, default_rights=RWXRWXRWX)
    clean_experiment_folder(flexp.get_file_path(""))
```
it will remove the log.txt as well, because it is now created by flexp.setup and not flog. Before calling `flexp.setup`, I don't know what is the experiment dir so I can't clean it.

I suggest adding a switch `clean_experiment_folder`, defaults to True, which removes everything in the experiment folder and only after that creates the log file.

What do you think?

Test is failing. Is it just me? But functionality is OK. 

Problem with test is cleaning data after test. 
`shutil.rmtree(expdir)`

```
(py3.7) tomas@tomas-Inspiron-5379:~/src/flexp/tests$ pytest test_flexp_override.py 
============================================================= test session starts =============================================================
platform linux -- Python 3.7.0, pytest-3.8.2, py-1.6.0, pluggy-0.7.1
rootdir: /media/tomas/data_storage/src/flexp, inifile:
collected 1 item                                                                                                                              

test_flexp_override.py F                                                                                                                [100%]

================================================================== FAILURES ===================================================================
________________________________________________________________ test_override ________________________________________________________________

    def test_override():
        # We have to reset the _eh to make flexp stop complaining about calling setup twice.
        flexp.core._eh = {}
        flexp.setup("tests/data/", "exp01", False, override_dir=False)
    
        expdir = path.join("tests/data/", "exp01")
        assert path.isdir(expdir), "flexp didn't create experiment dir with override_dir=False"
    
        # Test that it fails to create the directory, there should be logging file already.
        with pytest.raises(FileExistsError):
            flexp.core._eh = {}
            flexp.setup("tests/data/", "exp01", False, override_dir=False)
    
        # This should be ok
        flexp.core._eh = {}
        flexp.setup("tests/data/", "exp01", False, override_dir=True)
    
        # Remove the experiment dir
>       shutil.rmtree(expdir)

test_flexp_override.py:32: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
/home/tomas/.pyenv/versions/3.7.0/lib/python3.7/shutil.py:489: in rmtree
    onerror(os.rmdir, path, sys.exc_info())
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

path = 'tests/data/exp01', ignore_errors = False, onerror = <function rmtree.<locals>.onerror at 0x7f7718acde18>

    def rmtree(path, ignore_errors=False, onerror=None):
        """Recursively delete a directory tree.
    
        If ignore_errors is set, errors are ignored; otherwise, if onerror
        is set, it is called to handle the error with arguments (func,
        path, exc_info) where func is platform and implementation dependent;
        path is the argument to that function that caused it to fail; and
        exc_info is a tuple returned by sys.exc_info().  If ignore_errors
        is false and onerror is None, an exception is raised.
    
        """
        if ignore_errors:
            def onerror(*args):
                pass
        elif onerror is None:
            def onerror(*args):
                raise
        if _use_fd_functions:
            # While the unsafe rmtree works fine on bytes, the fd based does not.
            if isinstance(path, bytes):
                path = os.fsdecode(path)
            # Note: To guard against symlink races, we use the standard
            # lstat()/open()/fstat() trick.
            try:
                orig_st = os.lstat(path)
            except Exception:
                onerror(os.lstat, path, sys.exc_info())
                return
            try:
                fd = os.open(path, os.O_RDONLY)
            except Exception:
                onerror(os.lstat, path, sys.exc_info())
                return
            try:
                if os.path.samestat(orig_st, os.fstat(fd)):
                    _rmtree_safe_fd(fd, path, onerror)
                    try:
>                       os.rmdir(path)
E                       OSError: [Errno 39] Directory not empty: 'tests/data/exp01'

/home/tomas/.pyenv/versions/3.7.0/lib/python3.7/shutil.py:487: OSError
------------------------------------------------------------ Captured stderr call -------------------------------------------------------------
2018-10-05 20:06:22,761 - flexp.flexp.core - INFO - following files have been saved to experiment folder: /home/tomas/.pyenv/versions/py3.7/bin/pytest
-------------------------------------------------------------- Captured log call --------------------------------------------------------------
core.py                     82 WARNING  could not add group+write to experiment root dir tests/data/
core.py                    192 INFO     following files have been saved to experiment folder: /home/tomas/.pyenv/versions/py3.7/bin/pytest
========================================================== 1 failed in 0.08 seconds ===========================================================
```

Aha!
In Python 3.7, the previous test `test_working` fails:
```
$ pytest tests/
================================================================================= test session starts ==================================================================================
platform linux -- Python 3.7.0, pytest-3.8.2, py-1.6.0, pluggy-0.7.1
rootdir: /home/alan/tunnels/goedel/research/flexp, inifile:
collected 20 items                                                                                                                                                                     

tests/test_cache.py ...                                                                                                                                                          [ 15%]
tests/test_cache_attrs.py .                                                                                                                                                      [ 20%]
tests/test_chain.py .........                                                                                                                                                    [ 65%]
tests/test_flexp.py F.                                                                                                                                                           [ 75%]
tests/test_flexp_override.py F                                                                                                                                                   [ 80%]
tests/test_inspect.py ..                                                                                                                                                         [ 90%]
tests/test_parallel.py .                                                                                                                                                         [ 95%]
tests/test_utils.py .                                                                                                                                                            [100%]

======================================================================================= FAILURES =======================================================================================
_____________________________________________________________________________________ test_working _____________________________________________________________________________________

    def test_working():
        flexp.setup("tests/data/", "exp01", False)
    
        expdir = path.join("tests/data/", "exp01")
        assert path.isdir(expdir), "flexp didn't create experiment dir"
    
        with io.open("tests/testfile.txt", "wt") as x:
            x.write(u"hello")
    
        flexp.backup_files(["tests/testfile.txt"])
        assert path.exists(path.join("tests/data/", "exp01", "testfile.txt"))
    
        # forbid flexp creating metadata on exit
        if hasattr(atexit, "unregister"):
            getattr(atexit, "unregister")(flexp.core._eh["experiment"]._write_metadata)
    
        os.unlink("tests/testfile.txt")
        os.unlink(path.join(expdir, "testfile.txt"))
        if not hasattr(atexit, "unregister") and path.exists(
                path.join(expdir, "flexp_info.txt")):
            os.unlink(path.join(expdir, "flexp_info.txt"))
>       shutil.rmtree(expdir)

tests/test_flexp.py:33: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/home/alan/miniconda3/envs/test_flexp/lib/python3.7/shutil.py:489: in rmtree
    onerror(os.rmdir, path, sys.exc_info())
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

path = 'tests/data/exp01', ignore_errors = False, onerror = <function rmtree.<locals>.onerror at 0x7f301f2d1f28>

    def rmtree(path, ignore_errors=False, onerror=None):
        """Recursively delete a directory tree.
    
        If ignore_errors is set, errors are ignored; otherwise, if onerror
        is set, it is called to handle the error with arguments (func,
        path, exc_info) where func is platform and implementation dependent;
        path is the argument to that function that caused it to fail; and
        exc_info is a tuple returned by sys.exc_info().  If ignore_errors
        is false and onerror is None, an exception is raised.
    
        """
        if ignore_errors:
            def onerror(*args):
                pass
        elif onerror is None:
            def onerror(*args):
                raise
        if _use_fd_functions:
            # While the unsafe rmtree works fine on bytes, the fd based does not.
            if isinstance(path, bytes):
                path = os.fsdecode(path)
            # Note: To guard against symlink races, we use the standard
            # lstat()/open()/fstat() trick.
            try:
                orig_st = os.lstat(path)
            except Exception:
                onerror(os.lstat, path, sys.exc_info())
                return
            try:
                fd = os.open(path, os.O_RDONLY)
            except Exception:
                onerror(os.lstat, path, sys.exc_info())
                return
            try:
                if os.path.samestat(orig_st, os.fstat(fd)):
                    _rmtree_safe_fd(fd, path, onerror)
                    try:
>                       os.rmdir(path)
E                       OSError: [Errno 39] Directory not empty: 'tests/data/exp01'

/home/alan/miniconda3/envs/test_flexp/lib/python3.7/shutil.py:487: OSError
--------------------------------------------------------------------------------- Captured stderr call ---------------------------------------------------------------------------------
2018-10-09 09:04:12,284 - flexp.flexp.core - INFO - following files have been saved to experiment folder: tests/testfile.txt
---------------------------------------------------------------------------------- Captured log call -----------------------------------------------------------------------------------
core.py                    192 INFO     following files have been saved to experiment folder: tests/testfile.txt
____________________________________________________________________________________ test_override _____________________________________________________________________________________

    def test_override():
        # We have to reset the _eh to make flexp stop complaining about calling setup twice.
        flexp.core._eh = {}
>       flexp.setup("tests/data/", "exp01", False, override_dir=False)

tests/test_flexp_override.py:17: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
flexp/flexp/core.py:273: in setup
    clean_experiment_dir(override_dir)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

override_dir = False

    def clean_experiment_dir(override_dir):
        """
        Cleans experiment folder.
        :param bool override_dir: If true, clean the folder, if false, check it's empty or doesn't exists
        :return:
        """
        experiment_dir = get_file_path("")
        # Check experiment_dir is empty or doesn't exists
        if not override_dir:
            if not os.path.exists(experiment_dir) or not os.listdir(experiment_dir):
                return
            else:
>               raise FileExistsError("Directory {} exists and is not empty! Use override_dir=True to clean the directory.")
E               FileExistsError: Directory {} exists and is not empty! Use override_dir=True to clean the directory.

flexp/flexp/core.py:238: FileExistsError
=================================================================================== warnings summary ===================================================================================

```

--- a/flexp/flexp/core.py
+++ b/flexp/flexp/core.py
@@ -3,6 +3,7 @@ from __future__ import print_function
 from __future__ import absolute_import
 from __future__ import division
 
+import shutil
 from datetime import datetime
 from shutil import copyfile, copystat
 import io
@@ -222,8 +223,34 @@ class ExperimentHandler(object):
         self._metadata[key] = value
 
 
+def clean_experiment_dir(override_dir):
+    """
+    Cleans experiment folder.
+    :param bool override_dir: If true, clean the folder, if false, check it's empty or doesn't exists
+    :return:
+    """
+    experiment_dir = get_file_path("")
+    # Check experiment_dir is empty or doesn't exists
+    if not override_dir:
+        if not os.path.exists(experiment_dir) or not os.listdir(experiment_dir):
+            return
+        else:
+            raise FileExistsError("Directory {} exists and is not empty! Use override_dir=True to clean the directory.")
+    else:
+        # Dir doesn't exists, so everything's fine.
+        if not os.path.exists(experiment_dir):
+            return
+        # Remove all files in the directory
+        for the_file in os.listdir(experiment_dir):
+            file_path = os.path.join(experiment_dir, the_file)
+            if os.path.isfile(file_path):
+                os.unlink(file_path)
+            elif os.path.isdir(file_path):
+                shutil.rmtree(file_path)
+
+
 def setup(root_dir, exp_name=None, with_date=False, backup=True, default_rights=RWXRWS, loglevel=logging.DEBUG,
-          log_filename='log.txt', disable_stderr=False):
+          log_filename='log.txt', disable_stderr=False, override_dir=False):
     """Set up Experiment handler.
 
     :param root_dir: Path to experiments root directory. New directories will be created in the folder.
@@ -234,11 +261,17 @@ def setup(root_dir, exp_name=None, with_date=False, backup=True, default_rights=
     :param loglevel: Loglevel, DEBUG by default, can be overriden by FLEXP_LOGLEVEL environment variable.
     :param log_filename: Filename of log in experiment dir. 'log.txt' by default. If None, logging to file is disabled.
     :param disable_stderr: Enable/disable logging on stderr. False by default.
+    :param bool override_dir: If True, remove all contents of the directory.
+                              If False and there is any file in the directory, raise an exception.
     """
     if "experiment" in _eh:
         raise Exception("flexp.setup() called twice")
     _eh["experiment"] = ExperimentHandler(default_rights=default_rights)
     _eh["experiment"].setup(root_dir, exp_name, with_date=with_date)
+
+    # Clean the directory or check it's empty
+    clean_experiment_dir(override_dir)
+
     if backup:
         _eh["experiment"].backup_files([sys.argv[0]])
     atexit.register(_eh["experiment"]._write_metadata)

--- a/flexp/flexp/core.py
+++ b/flexp/flexp/core.py
@@ -364,4 +364,5 @@ def set_metadata(key, value):
 def disable():
     """Surpress all output and returns /dev/null as path to all files."""
     _eh["experiment"].disabled = True
+    _setup_logging(logging.FATAL, None, True)
 

--- a/flexp/flexp/core.py
+++ b/flexp/flexp/core.py
@@ -235,7 +235,11 @@ def clean_experiment_dir(override_dir):
         if not os.path.exists(experiment_dir) or not os.listdir(experiment_dir):
             return
         else:
-            raise FileExistsError("Directory {} exists and is not empty! Use override_dir=True to clean the directory.")
+            raise FileExistsError(
+                "Directory '{}' already exists and is not empty! "
+                "Use override_dir=True to clean the directory."
+                    .format(name())
+            )
     else:
         # Dir doesn't exists, so everything's fine.
         if not os.path.exists(experiment_dir):

--- a/flexp/flexp/logging.py
+++ b/flexp/flexp/logging.py
@@ -5,8 +5,6 @@ import six
 
 log_formatter = logging.Formatter('%(asctime)s - %(name)s - %(levelname)s - %(message)s')
 
-# file_handler_global = []
-
 
 def loglevel_from_string(level):
     """
@@ -60,4 +58,3 @@ def _close_file_handlers():
     root_logger = logging.getLogger()
     for file_handler in root_logger.handlers:
         file_handler.close()
-

--- a/flexp/flexp/logging.py
+++ b/flexp/flexp/logging.py
@@ -3,8 +3,9 @@ import warnings
 import os
 import six
 
-
 log_formatter = logging.Formatter('%(asctime)s - %(name)s - %(levelname)s - %(message)s')
+
+# file_handler_global = []
 
 
 def loglevel_from_string(level):
@@ -39,6 +40,7 @@ def get_loglevel_from_env(default_level):
 
 
 def _setup_logging(level=logging.DEBUG, filename='log.txt', disable_stderr=False):
+    _close_file_handlers()
     level = loglevel_from_string(level)
     root_logger = logging.getLogger()
     root_logger.setLevel(level)
@@ -52,3 +54,10 @@ def _setup_logging(level=logging.DEBUG, filename='log.txt', disable_stderr=False
         root_logger.addHandler(stream_handler)
 
     warnings.simplefilter("once")
+
+
+def _close_file_handlers():
+    root_logger = logging.getLogger()
+    for file_handler in root_logger.handlers:
+        file_handler.close()
+

--- a/tests/test_flexp.py
+++ b/tests/test_flexp.py
@@ -30,6 +30,7 @@ def test_working():
     if not hasattr(atexit, "unregister") and path.exists(
             path.join(expdir, "flexp_info.txt")):
         os.unlink(path.join(expdir, "flexp_info.txt"))
+    flexp.disable()
     shutil.rmtree(expdir)
 
 

--- a/tests/test_flexp_override.py
+++ b/tests/test_flexp_override.py
@@ -12,11 +12,16 @@ from flexp import flexp
 
 
 def test_override():
+    expdir = path.join("tests/data/", "exp01")
+
+    # Remove the experiment dir if it exists
+    if os.path.exists(expdir):
+        shutil.rmtree(expdir)
+
     # We have to reset the _eh to make flexp stop complaining about calling setup twice.
     flexp.core._eh = {}
     flexp.setup("tests/data/", "exp01", False, override_dir=False)
 
-    expdir = path.join("tests/data/", "exp01")
     assert path.isdir(expdir), "flexp didn't create experiment dir with override_dir=False"
 
     # Test that it fails to create the directory, there should be logging file already.
@@ -28,6 +33,9 @@ def test_override():
     flexp.core._eh = {}
     flexp.setup("tests/data/", "exp01", False, override_dir=True)
 
-    # Remove the experiment dir
-    shutil.rmtree(expdir)
+    # Disable logging to be able to delete the experiment directory.
+    flexp.disable()
 
+    # Remove the experiment dir
+    if os.path.exists(expdir):
+        shutil.rmtree(expdir)

--- a/tests/test_flexp_override.py
+++ b/tests/test_flexp_override.py
@@ -1,0 +1,33 @@
+# coding: utf-8
+
+import atexit
+import io
+import os
+from os import path
+import shutil
+
+import pytest
+
+from flexp import flexp
+
+
+def test_override():
+    # We have to reset the _eh to make flexp stop complaining about calling setup twice.
+    flexp.core._eh = {}
+    flexp.setup("tests/data/", "exp01", False, override_dir=False)
+
+    expdir = path.join("tests/data/", "exp01")
+    assert path.isdir(expdir), "flexp didn't create experiment dir with override_dir=False"
+
+    # Test that it fails to create the directory, there should be logging file already.
+    with pytest.raises(FileExistsError):
+        flexp.core._eh = {}
+        flexp.setup("tests/data/", "exp01", False, override_dir=False)
+
+    # This should be ok
+    flexp.core._eh = {}
+    flexp.setup("tests/data/", "exp01", False, override_dir=True)
+
+    # Remove the experiment dir
+    shutil.rmtree(expdir)
+

